Pass information for ongoing press events in Matter

phoffmeister opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
~This increases the `max_pressed_supported` to 18, which is needed for BILRESA.~
~Additionally this also adds `multi_press_ongoing` to the supported event types, which then allows them to propagate.~

### EDIT
Original code 
* ignore `multi_press_ongoing`
* map `multi_press_complete` together with capped `totalNumberOfPressesCounted` to custom events `multi_press_1` ... `multi_press_n`

Proposed change
* pass `multi_press_ongoing` and `multi_press_complete` transparently
* add convenience event types for common scenarios `event_type_extra`, with values `single`, `double` and `triple`

This removes the limit of multi presses on HA side and transparently passes available events, while still having the convenience for simple cases.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #159035
- This PR is related to issue: #159035
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [ ] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

phoffmeister committed:
pass information for ongoing press events

phoffmeister committed:
add `multi_press_ongoing`

TheJulianJES reviewed (commented):
If we add the `multi_press_ongoing` state to the event entity, we should add a translation for that. It should be enough to add it to `strings.json`, since it's already in `EVENT_TYPES_MAP`. So just needs to be added here: https://github.com/home-assistant/core/blob/f786ec18a9a39661d58665c5cc919cb9ac934a9f/homeassistant/components/matter/strings.json#L150-L173

---

I do think we should carefully consider adding back the "multi press ongoing" event, as it was explicitly removed in https://github.com/home-assistant/core/pull/122191.

It makes total sense for the IKEA dimmer, but just not sure if there was more background to removing it, other than the "We ignore multi press ongoing as it doesn't make a lot of sense and many devices do not support it" reason from the comment. I'll try to get some more info on that last part.

pitxinbio reviewed (commented):
This now should be included for Ikea Bilresa wheel dimmer and probably others as it is reporting matter events as ongoing multi state up to 18. This will allow to dim the lights and see how the lamps react in real time while scrolling. Now this makes sense to consider ongoing multistates and maximize the limit to 18.

phoffmeister committed:
Merge branch 'dev' into matter/event-multi-press-ongoing

phoffmeister commented:
> > Maybe it makes sense to remove `multi_press_{}` events entirely as this is/could be reported with either `multi_press_ongoing` + `currentNumberOfPressesCounted` or `multi_press_complete` + `totalNumberOfPressesCounted`
> 
> After looking at [#122191](https://github.com/home-assistant/core/pull/122191) it seems they were implemented to ease the use in automations and explicitly not having to check the `totalNumberOfPressedCounted` variable. I kinda see the point, you can just check for the event without the need for checking the count variable. Also removing them would be a quite breaking change, right?

Yes, agree. It's very convenient to have these states explicit.
On the other hand this will grow to a big list of events.
```
event_types:
        - multi_press_ongoing
        - multi_press_1
        - multi_press_2
        - multi_press_3
        - multi_press_4
        - multi_press_5
        - multi_press_6
        - multi_press_7
        - multi_press_8
        - multi_press_9
        - multi_press_10
        - multi_press_11
        - multi_press_12
        - multi_press_13
        - multi_press_14
        - multi_press_15
        - multi_press_16
        - multi_press_17
        - multi_press_18
```
Which I think could be avoided. 

I think the most important ones are `multi_press_{1,2,3}` as they represent more realistic scenarios for "normal" buttons.
Maybe we could handle those cases by adding some extra data before passing it along.
This way we would pass the event_types transparently and give the extra convenience for the single, double and tripple clicks.

```python
    @callback
    def _on_matter_node_event(
        self,
        event: EventType,
        data: MatterNodeEvent,
    ) -> None:
        """Call on NodeEvent."""
        if data.endpoint_id != self._endpoint.endpoint_id:
            return
        event_type = EVENT_TYPES_MAP[data.event_id]

        if event_type == "multi_press_complete":
            if data.data:
                presses = data.data.get("totalNumberOfPressesCounted", 1)
                if presses == 1:
                    data.data['event_type_extra'] = "single"
                elif presses == 2:
                    data.data['event_type_extra'] = "double"
                elif presses == 3:
                    data.data['event_type_extra'] = "tripple"

        if event_type not in self.event_types:
            # this should not happen, but guard for bad things
            # some remotes send events that they do not report as supported (sigh...)
            return

        # pass the rest of the data as-is (such as the advanced Position data)
        self._trigger_event(event_type, data.data)
        self.async_write_ha_state()

```

with the event types like this
```
event_types:
        - multi_press_ongoing
        - multi_press_complete
```
and the newly added
```
event_type_extra:
        - single
        - double
        - tripple
```

Yes, this would break existing automations.

TheGreyDiamond commented:
This makes the most sense to me. Have a few "standard" amounts (once, twice, triple) and expose the rest as multipress with a counter. This allows for easy automations for simple actions while enabling more complex rules for those who need it.

phoffmeister committed:
pass ongoing and complete events

For convenience also pass `event_type_extra` information.
Possible values are `single`, `double` and `triple` for their
corresponding click actions.

phoffmeister committed:
test event_type_extra is passed correctly

phoffmeister committed:
update multi-press event labels to ongoing/complete

phoffmeister committed:
Merge branch 'dev' into matter/event-multi-press-ongoing

MrEbbinghaus commented:
@phoffmeister 
Both your ideas are good. Translating the event to the old 1-8 one feels like the best way, but it's a compromise. The alternative is a breaking change, it is a tricky situation. :-/

I prefer your new event. Keeping the number of presses as structured metadata instead of munging it into the name is great!

Is there precedent on how other integrations handle this? Hue has `event_types: initial_press, repeat, short_release, long_press, long_release`. Something like this should be consistent across HA, otherwise you have to change everything when you replace e.g. a Hue switch with a Matter switch.

3oris commented:
@MrEbbinghaus @phoffmeister 

Removing the multipress support and replacing it by the limited count support we have now was a mistake in the first place. AND it was a breaking change at that time as well. So of course reverting that should be expected to be a breaking change. I still believe it would be best to get to a clean multipress support implementation (AND NOT a hybrid solution that counts presses up to 8 and only then switches over to multipress).
Yes that would be breaking, and even though it would affect me as well, I still think it would be better to have it _now_ than keep collecting users, then realizing that the hybrid approach was a foul compromise, and then being forced to still do a breaking change on a potentially way larger user base (now that real devices are being deployed to the mass market)...
_Now_ is possibly the last point in time to have a breaking change with only limited bad impact.

Just my 2ct

Waghead commented:
> @MrEbbinghaus @phoffmeister
> 
> Removing the ongoing multipress support and replacing it by the limited count support we have now was a mistake in the first place. AND it was a breaking change at that time as well. So of course reverting that should be expected to be a breaking change. I still believe it would be best to get to a clean ongoing multipress support implementation (AND NOT a hybrid solution that counts presses up to 8 and only then switches over to ongoing multipress). Yes that would be breaking, and even though it would affect me as well, I still think it would be better to have it _now_ than keep collecting users, then realizing that the hybrid approach was a foul compromise, and then being forced to still do a breaking change on a potentially way larger user base (now that real devices are being deployed to the mass market)... _Now_ is possibly the last point in time to have a breaking change with only limited bad impact.
> 
> Just my 2ct, a _users_ point of view.

This is the way

phoffmeister committed:
refactor multi press names to map

phoffmeister committed:
Merge branch 'matter/event-multi-press-ongoing' of github.com:phoffmeister/core into matter/event-multi-press-ongoing

phoffmeister commented on homeassistant/components/matter/event.py:
```diff
@@ -117,12 +109,17 @@ def _on_matter_node_event(
+        if event_type == "multi_press_complete":
+            if data.data:
+                presses = data.data.get("totalNumberOfPressesCounted", 1)
+                if presses == 1:
```
I opted for the map solution.

phoffmeister committed:
Merge branch 'dev' into matter/event-multi-press-ongoing

phoffmeister commented:
@3oris Thank you for the extra context.
Code is ready from my side.

GTx220 commented:
Breaking change will affect me too (although not much) however, I do believe its the right way to do it. No point moving forward with a compromise especially as we're likely to see more devices over time that would work better this way.

Gidobrinkhof commented:
Why is this all needed when the (by default disabled by HA) current switch state of the 9 actions can be used? They give a pulse when rotating each segment a specific direction for each page.. I managed to get the real time updates since release without any of this multi_press nonsense!

apocer commented:
@Gidobrinkhof I have the IKEA Bilresa remote and only get events when I stop turning the wheel. No events while turning the wheel, which is crucial for dimming lights. `multi_press_ongoing` is being sent from the matter network, but the current implementation supresses this event. IKEA did not implement the scroll wheel very well, but HA should also not suppress events that have potential value for automations.

marcin-igel commented:
@Gidobrinkhof 
Enabling those diag entities is not a solution to addressing how these devices are meant to be used, and building automations based on arcane diag entities is a clunky solution that only applies to this specific device. As it is now - this PR is a refactor of a solution that was initially implemented without enough foresight, noone's fault, and provides a way to utilize all matter devices that provide feedback as ongoing multipress.

The fact that you can use those 9 diag entities for building automations using that particular device is not a valid argument against replacing existing implementation with future-proof one that's aligned more to matter-standard way of communicating events by devices.

The goal here is to make sure that all HA users of matter devices now and **in the future especially** have the most painless way of using their devices, and have the best possible experience.

As stated - this will be a breaking change, but it's absolutely necessary. Without it we accept the debt the current incomplete support creates. Hope that helps with the bigger picture.

philiplb commented:
Any chance this will make it into 2026-02? 🫣

mavanmanen commented:
A decision needs to be made, clearly a lot of people are waiting on this and considering the popularity of the ikea remotes it will only increase.

MrEbbinghaus reviewed (approved):
@philiplb My _requested changes_ were that we think/talk about the breaking change. 
But I think @3oris summarised it nicely in https://github.com/home-assistant/core/pull/159045#issuecomment-3708097664.

I'm fine with breaking changes, when they are done thoughtfully. And there is long term gain in a clean system.

WEZANGO commented:
Even after this PR is merged, scrolling still won't be smooth, am I right? Does this still require proper rotary implementation on Matter side?

MrEbbinghaus committed:
Allow events based on feature map alone

MrEbbinghaus commented on homeassistant/components/matter/event.py:
```suggestion

        # Either momentary xor latching is required
        if feature_map & SwitchFeature.kLatchingSwitch:
            event_types.append("switch_latched")

        if feature_map & SwitchFeature.kMomentarySwitch:
            event_types.append("initial_press")

        # The specs are more strict about what features a device must support
        # We just trust the device. No harm in expecting more events.

        # release is optional, but requires momentary support
        if feature_map & SwitchFeature.kMomentarySwitchRelease:
            event_types.append("short_release")

        # multi press is optional, but requires release support
        if feature_map & SwitchFeature.kMomentarySwitchMultiPress:
            event_types.append("multi_press_ongoing")
            event_types.append("multi_press_complete")

        # long press is optional, but requires release support
        if feature_map & SwitchFeature.kMomentarySwitchLongPress:
            event_types.append("long_press")
            event_types.append("long_release")
```

phoffmeister replied:
I initially didnt want to introduce too many changes at once to keep the PR focused. But Im of course fine to also include the changes about responsiveness.

MrEbbinghaus committed:
Update snapshots

MrEbbinghaus committed:
Replace hardcoded ids with ids from matter lib

MrEbbinghaus commented:
I checked the matter server logs. It looks like the Bilresa doesn't emit Events for every 'click'. 

I turned the wheel quickly and this is one of the captures: 


|    Time | Event                   |
|--------:|:-------------------------|
|     0ms | InitialPress            |
|    +5ms | + ShortRelease          |
|  +510ms | InitialPress            |
|  +545ms | MultiPressOngoing(14)   |
|  +549ms | + ShortRelease          |
| +1009ms | InitialPress            |
| +1013ms | + MultiPressOngoing(18) |
| +1016ms | + ShortRelease          |
| +1509ms | MultiPressComplete(18)  |
| +1610ms | InitialPress            |
| +1613ms | + ShortRelease          |
| +2110ms | InitialPress            |
| +2115ms | + MultiPressOngoing(14) |
| +2120ms | + ShortRelease          |
| +2610ms | InitialPress            |
| +2648ms | MultiPressOngoing(18)   |

(Events with + are sent in the same report as the event before it.)

phoffmeister committed:
Merge pull request #1 from MrEbbinghaus/matter/event-multi-press-ongoing-ebbinghaus

Allow events based on feature map alone

phoffmeister committed:
Merge branch 'dev' into matter/event-multi-press-ongoing

phoffmeister committed:
update snapshots

phoffmeister committed:
Merge branch 'dev' into matter/event-multi-press-ongoing

phoffmeister committed:
Merge branch 'dev' into matter/event-multi-press-ongoing

phoffmeister committed:
Merge branch 'dev' into matter/event-multi-press-ongoing

phoffmeister committed:
Merge branch 'dev' into matter/event-multi-press-ongoing

etillxd commented:
> 
> Edit: I'm just realizing the Matter integration does not expose the `Step` command so with Matter lights the behaviour may not be that good as it depends on the brightness level known by HA, whereas the step command is stateless. For the perfect replica of IKEA dimming the native Step is needed.
> 

The step command would basically tell the light "change brightness by 1” (so, a relative change), but right now we can only send an absolute value, right?
So the current flow is that we take the last known brightness, add 1 (or whatever step) on our side, send that as a new absolute value, and then wait for the light to report its updated state before we can change it again. Did I understand that correctly?
Would it be possible to implement the step command in Home Assistant?

phoffmeister commented on homeassistant/components/matter/strings.json:
```diff
@@ -159,14 +159,8 @@
-              "multi_press_6": "Pressed 6 times",
-              "multi_press_7": "Pressed 7 times",
-              "multi_press_8": "Pressed 8 times",
+              "multi_press_complete": "Multiple presses finished",
```
@MrEbbinghaus  We would lose the generic `multi_press_complete` from the dropdown in the automations, right?

phoffmeister replied:
I managed to get it working briefly, though I unfortunately didn’t take a screenshot.

I liked the suggestion about the `translation_placeholders` changes, and I agree there’s still room for improvement. I think it’s best to address that in a follow up PR since there also seems to be an issue with the translations and the dropdown in automations.

Otherwise this PR might grow too much in scope.

arturpragacz reviewed (changes requested):
Preserve the old entity to not break backwards compatibility and deprecate it to give time to users. For the new behaviour use a new entity instead.

Also put the number of presses in the emitted event data.

MartinHjelmare commented:
There are requested changes that haven't been addressed. This PR is waiting for that.

StephanMeijer commented:
Hey! We've been working on a BILRESA scroll wheel dimmer for Home Assistant and independently arrived at the same conclusions as this PR. We've published a complete working solution as a gist: https://gist.github.com/StephanMeijer/75a26615f9b692cf7ce7cf8d6716ade4

### Our empirical findings

We instrumented `device_controller.py` inside `addon_core_matter_server` with `SWITCH_EVENT` debug logging to measure exact event timing from the BILRESA scroll wheel. Key observations:

- **Events arrive individually every ~500ms** — this is the Matter MultiPress window timeout hardcoded in BILRESA firmware. There is no multi-second SED batching.
- **Transport delay**: 30–150ms from device to HA.
- **Each detent**: `InitialPress → MultiPressOngoing(count=N) → ShortRelease` in ~40ms, then 500ms pause before the next detent.
- **Fast scroll**: `currentNumberOfPressesCounted` jumps to 18 (`MultiPressMax`) within a single 500ms window.
- **Slow scroll**: count increments gradually (1, 2, 3...).
- **CW/CCW endpoints can fire within 1ms** of each other during direction transitions.

### What we built on top of the patch

Using the exposed `multi_press_ongoing` events and `currentNumberOfPressesCounted`, we built a quadratic acceleration dimmer:

```
step = clamp(count² ÷ 6, 2, 50)
```

- Slow rotation → step=2 (fine adjustment)
- Fast spin (count=18) → step=50 (~2.5s full range sweep)

The automation uses `input_number` helpers for own-state brightness tracking (avoids stale Matter state during rapid scrolling), direction locking (prevents bounce during CW↔CCW transitions), and a sync automation with a 2s idle guard.

### Bugs found along the way

1. **`input_number` max=254**: HA's `input_number` entity caps at 254. Setting brightness to 255 causes a **silent automation crash** — the automation enters error state and stops all future triggers. Documented in the gist.
2. **Jinja2 scoping**: Variables set inside `{% if %}` blocks aren't visible outside. Requires top-level `{% set %}`.
3. **Direction bounce**: CW/CCW endpoints firing within 1ms needs a 400ms direction lock with an explicit `last_dir == 0` (idle) case.

### How our patch aligns with this PR

Our v4 patch (in the gist) now matches this PR's approach:

- Flat `if` chain instead of `elif` nesting — "We just trust the device"
- `multi_press_ongoing` + `multi_press_complete` as event types instead of `multi_press_1..N` enumeration
- `MULTI_PRESS_COUNT_TO_NAME` dict with `event_type_extra` field (`"single"`, `"double"`, `"triple"`)
- Simplified `_on_matter_node_event` — single `EVENT_TYPES_MAP` lookup

Our earlier patch versions independently discovered the same `elif` bug (v1: `kMomentarySwitchMultiPress` checked before `kMomentarySwitch`, silently dropping `initial_press`).

### What differs

The only difference is that our patch doesn't address the backwards compatibility concern raised by @arturpragacz (old entity preservation + deprecation path). For users patching locally that's irrelevant — but it's obviously the right thing for upstream.

Hope the timing data and real-world scroll wheel use case helps move this forward. The BILRESA scroll wheel is genuinely great hardware that's held back by the integration not exposing these events.

---

_Claude Opus 4_

etillxd commented:
It looks like there hasn't really been any activity on this PR for a few months now, especially after the requested changes. @phoffmeister are you still interested in working on this, or has there been some activity we can't see here?

utkuturunc commented:
i gave it a try on doing the updates on this pr: https://github.com/phoffmeister/core/pull/2/changes

Fraeco commented:
> i gave it a try on doing the updates on this pr: https://github.com/phoffmeister/core/pull/2/changes

There is also [this PR](https://github.com/home-assistant/core/pull/168367) by @StephanMeijer that implements this change with the requested backwards compatibility. 
Maybe someone from the project could take a look at both these PR's and decide if they want to move on with either one of these
